### DISCUSSION TITLE
installation command to be updated?

### DISCUSSION THREAD

**Discussion by giovannicoppola** [mkb79/Audible#97] [General]
Hi @mkb79 I hope you are well! A user of my workflow could not install the package and I realized my command fails now (something changed?). 
This below is my requirement.txt file, should I change anything? 

`pyparsing<3,>=2.4.2`
`audible-cli @ git+https://github.com/mkb79/audible-cli@development`

Thanks!

**Comment by mkb79** [comment]
Hi @giovannicoppola,

I'm doing well so far. The job is just stressful right now. 

There is an issue in your requirements.txt. You have to change `audible-cli @ git+https://github.com/mkb79/audible-cli@development` to `audible-cli`. The reason for this is the deletion of the development branch after merging into master and releasing v0.2.0!

**Comment by giovannicoppola** [comment]
awesome thanks! I get this error but the library seems to be there? (not sure how to simulate a fresh install on my machine)

`ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.`
`osxmetadata 0.99.14 requires pyobjc-core==7.2, but you have pyobjc-core 8.5 which is incompatible.`
`osxmetadata 0.99.14 requires pyobjc-framework-AppleScriptKit==7.2, but you have pyobjc-framework-applescriptkit 8.5 which is incompatible.`
`osxmetadata 0.99.14 requires pyobjc-framework-AppleScriptObjC==7.2, but you have pyobjc-framework-applescriptobjc 8.5 which is incompatible.`

**Reply by mkb79** [reply]
@giovannicoppola 

Update `osxmetadata` to the latest version (v0.99.37) by executing `pip install --upgrade osxmetadata` in your shell! This version support `pyobjc-framework-AppleScriptObjC 8`

**Reply by giovannicoppola** [reply]
thanks! should I add a minimum version of osxmetadata in my requirements.txt file?

**Reply by mkb79** [reply]
If you need osxmetadata for your project, then yes! If you have it installed for other purposes, then no ;)!

**Reply by giovannicoppola** [reply]
oh I see, pip just checks all the packages, and I have an older `osxmetadata`? but `osxmetadata` is not required for your project?

**Reply by mkb79** [reply]
Yeah, your version of `osxmetadata` is outdated. `audible-cli` doesn’t require this package.

**Comment by mkb79** [comment]
@giovannicoppola 
Do you want to implement an Alfred workflow for goodreads?

**Reply by giovannicoppola** [reply]
if you are creating a package, sure! There is [one](https://www.alfredforum.com/topic/14296-goodreads-book-search/#comment-73190), but it needs to be updated to Python 3, and will become obsolete as goodreads [plans to disable](https://joealcorn.co.uk/blog/2020/goodreads-retiring-API) the API keys (if they haven't already)

**Reply by mkb79** [reply]
[Here](https://gist.github.com/mkb79/2a3f0acb3a75d3cf82d0c85a29497d22) you can find a gist from me which let you register a Goodreads iOS device. After registration you obtain for e.g. an access token. Then you can request the [Goodreads API](https://www.goodreads.com/api) using x-amz-access-token ACCESS_TOKEN and User-Agent Goodreads/4.0.1 (iPhone; iOS 15.4.1; Scale/3.00) in your request headers.

These doesn’t need a Goodreads API developer key. And there are some undocumented API endpoints too.

**Reply by giovannicoppola** [reply]
awesome, I'd love to work on that! as for audible, I would like to 1) query my library 2) query the main library 3) be able to assign or remove books from the catalog to my library 4) have some information about books in my library (if read, author, shelf etc). 

Are these possible with the endpoints that are currently available?

**Reply by mkb79** [reply]
[Here](https://www.goodreads.com/api) is a description for all public API endpoints. 

And below are some undocumented:

```
https://www.goodreads.com/api/current_user_data?_nc=true&currently_reading=true&format=xml&id={USER_ID}&include_social_shelving_info=true&user_shelves=true&v=2

https://www.goodreads.com/api/v3/updates/newsfeed?_nc=true&format=xml&max_updates=10

https://www.goodreads.com/api/book/basic_book_data/182506?_extras%5Bbook_covers_large%5D=true&_nc=true&format=xml

https://www.goodreads.com/api/book/book_reviews/182506?_extras%5Bbook_covers_large%5D=true&_nc=true&format=xml

https://www.goodreads.com/api/book/related_books/182506?_extras%5Bbook_covers_large%5D=true&_nc=true&format=xml

https://www.goodreads.com/api/gca_metadata?_nc=true&format=xml&locale=de-DE

https://www.goodreads.com/api/current_user_shelves?_nc=true&format=xml
```

I think you can do all what you want!

**Comment by giovannicoppola** [comment]
Hi @mkb79,
I finally started to work on this! I suspect I will need some help and I
hope I can ask you a few questions. What is the right forum/discussion?
First question: as I am testing the register function, I realized I am
registering a new device every time. Is there a way to delete them?
Thanks!

**Comment by mkb79** [comment]
> First question: as I am testing the register function, I realized I am
registering a new device every time. Is there a way to delete them?

Hi @giovannicoppola,
You can deregister a device using an `Authenticator` instance with your auth data like so:

```python
import audible
auth = Authenticator.from_file(…)
auth.deregister_device()
```

After a successfully deregistration you can delete the auth file. You can use the Amazon website to remove a device, too.

**Comment by giovannicoppola** [comment]
sorry the thread might have been lost, I was referring to your Goodreads
API script?

**Reply by mkb79** [reply]
Take a look on my [GoodreadsAPi gist](https://gist.github.com/mkb79/2a3f0acb3a75d3cf82d0c85a29497d22). There you can find a function to deregister a Goodreads device. Please take attention that the access token is not expired. Otherwise you have to refresh them before you deregister.

**Comment by giovannicoppola** [comment]
Thanks, still having trouble interacting with the API. Would you have an
example of a GET request in Python (I am using the `requests` package) to
get, for example, a user's shelves
<https://www.goodreads.com/api/index#shelves.list>?

**Reply by mkb79** [reply]
You have to authenticate to the API using your access token, retrieved after a device registration or a token refresh request (if the access token was expired)! Then you have to add the access token to your request as the `x-amz-access-token` header. Then you can request all API endpoints from https://www.goodreads.com/api/index without providing the developer key. 

I've wrote an private POC package. This uses the `httpx` package. I can share this with you using a private channel (e.g. mail).

**Reply by mkb79** [reply]
This should work with requests module and a valid access token:

```python
import requests

access_token = ...
base_headers = {
    "User-Agent": "Goodreads/4.0.1 (iPhone; iOS 15.4.1; Scale/3.00)",
    "Accept-Language": "en-US;q=1",
    "x-amz-access-token": access_token
}

with requests.Session(headers=base_headers) as client:
    # get user id
    r = client.get("https://www.goodreads.com/api/auth_user")
    user_id = ...
    
    params = {"user_id": user_id}
    r = client.get("https://www.goodreads.com/shelf/list.xml", params=params)
```

**Reply by giovannicoppola** [reply]
no luck 😟

**Reply by mkb79** [reply]
Are you able to refresh your Goodreads access token with your refresh token? I tried it out right now and all working fine for me (refreshing, get user shelves and user infos from API)!

**Reply by mkb79** [reply]
@giovannicoppola 
I've invited you to my goodreads private repo. You can use it but status is pre-alpha. There we can discuss more how to use.

**Comment by giovannicoppola** [comment]
I get a 401 code using this below. my email is my username AT [gmail.com](http://gmail.com/). thanks!!
```
r=requests.get("https://www.goodreads.com/shelf/list.xml", 
    headers={
        "x-amz-access-token":ACCESS_TOKEN, 
        "User-Agent": 'Goodreads/4.0.1 (iPhone; iOS 15.4.1; Scale/3.00)'
        })
```

**Reply by mkb79** [reply]
Look on my example above for using the requests module. Is your access token is valid? You don’t provide a user_id. Please try this endpoint `https://www.goodreads.com/api/current_user_shelves?_nc=true&format=xml`. This should work without an user_id.

**Comment by giovannicoppola** [comment]
it works if I create a new device, so you are right it must have to do with
the refresh token. is 3600 in `'expires_in': '3600'` seconds?

**Comment by mkb79** [comment]
Yeah, an access token expires after 1 hour. Then you need to refresh the token.

**Reply by mkb79** [reply]
You don’t need a new device registration for refreshing the access token. You can use the refresh token for this.

**Reply by giovannicoppola** [reply]
yes, I understand now, thanks!! 

also, I don't seem to be able to pass headers to `requests.Session()` (error below)... 
```
with requests.Session(headers=base_headers) as client:
TypeError: __init__() got an unexpected keyword argument 'headers'
```
... but it works if I add it to `client.get`:  `r = client.get("https://www.goodreads.com/api/auth_user",headers=base_headers)`

**Reply by mkb79** [reply]
Okay. I thought `requests` supports `headers` on instantiating. But I'm wrong. Sorry for that!

**Comment by giovannicoppola** [comment]
no luck 😟

**Reply by mkb79** [reply]
No luck with what?

**Reply by giovannicoppola** [reply]
something weird is happening, this message and the one after it are when we had this conversation, 12 days ago. not sure why they got resent. all fixed for now!

**Comment by giovannicoppola** [comment]
I get a 401 code using this below. my email is my username AT gmail.com.
thanks!!
r=requests.get("https://www.goodreads.com/shelf/list.xml",
headers={
"x-amz-access-token":ACCESS_TOKEN,
"User-Agent": 'Goodreads/4.0.1 (iPhone; iOS 15.4.1; Scale/3.00)'
})
